### ISSUE TITLE
exo package buck install SocketTimeoutException

### ISSUE THREAD

**Issue by philrabin** [facebook/buck#1369]
Getting a SocketTimeoutException. This happens on emulators and devices of multiple Android versions.

Buck version 800d5a9669417d5a8b4bc455e34a629422ed1ff1

```
[-] PROCESSING BUCK FILES...FINISHED 1.2s
[-] DOWNLOADING... (0.00 B/S AVG, TOTAL: 0.00 B, 0 Artifacts)
[-] BUILDING...FINISHED 0.4s [100%] (2/2 JOBS, 0 UPDATED, 0 [0.0%] CACHE MISS)
    Details: http://localhost:51735/trace/4e7e1fcc-add2-4270-9a9d-988e1dc4ddac
[+] INSTALLING...25.2s
BUILD FAILED: Failed: install exopackage apk
java.util.concurrent.ExecutionException: java.lang.RuntimeException: com.facebook.buck.android.AdbHelper$CommandFailedException: Command 'umask 022 && dalvikvm -classpath /data/app/com.facebook.buck.android.agent-1.apk:/data/app/com.facebook.buck.android.agent-2.apk:/data/app/com.facebook.buck.android.agent-1/base.apk:/data/app/com.facebook.buck.android.agent-2/base.apk com.facebook.buck.android.agent.AgentMain receive-file 2828 3296974 /data/local/tmp/exopackage/com.ubercab.presidio.exo/secondary-dex/secondary-33af3484b3dc1dc29030b4e5680412ae7202d670.dex.jar ; echo -n :$?' failed with code 1.  Output:
53616f791d97484537820a0d445f3965Jun 07, 2017 5:43:01 PM com.facebook.buck.android.agent.AgentMain main
SEVERE: Command failed
java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:151)
	at java.net.SocketInputStream.read(SocketInputStream.java:120)
	at java.net.SocketInputStream.read(SocketInputStream.java:106)
	at com.facebook.buck.android.agent.AgentMain.receiveAndValidateSessionKey(AgentMain.java:186)
	at com.facebook.buck.android.agent.AgentMain.doReceiveFile(AgentMain.java:147)
	at com.facebook.buck.android.agent.AgentMain.main(AgentMain.java:70)


	at com.google.common.util.concurrent.AbstractFuture.getDoneValue(AbstractFuture.java:500)
	at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:479)
	at com.google.common.util.concurrent.AbstractFuture$TrustedFuture.get(AbstractFuture.java:76)
	at com.facebook.buck.android.AdbHelper.adbCall(AdbHelper.java:374)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$RealAdbInterface.adbCall(ExopackageInstaller.java:139)
	at com.facebook.buck.android.exopackage.ExopackageInstaller.install(ExopackageInstaller.java:204)
	at com.facebook.buck.android.AdbHelper.installApk(AdbHelper.java:563)
	at com.facebook.buck.cli.InstallCommand.installApk(InstallCommand.java:364)
	at com.facebook.buck.cli.InstallCommand.install(InstallCommand.java:243)
	at com.facebook.buck.cli.InstallCommand.runWithoutHelp(InstallCommand.java:211)
	at com.facebook.buck.cli.AbstractCommand.run(AbstractCommand.java:222)
	at com.facebook.buck.cli.AbstractContainerCommand.run(AbstractContainerCommand.java:66)
	at com.facebook.buck.cli.BuckCommand.run(BuckCommand.java:80)
	at com.facebook.buck.cli.Main.runMainWithExitCode(Main.java:900)
	at com.facebook.buck.cli.Main.runMainThenExit(Main.java:329)
	at com.facebook.buck.cli.Main.nailMain(Main.java:1674)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.martiansoftware.nailgun.NGSession.run(NGSession.java:338)
Caused by: java.lang.RuntimeException: com.facebook.buck.android.AdbHelper$CommandFailedException: Command 'umask 022 && dalvikvm -classpath /data/app/com.facebook.buck.android.agent-1.apk:/data/app/com.facebook.buck.android.agent-2.apk:/data/app/com.facebook.buck.android.agent-1/base.apk:/data/app/com.facebook.buck.android.agent-2/base.apk com.facebook.buck.android.agent.AgentMain receive-file 2828 3296974 /data/local/tmp/exopackage/com.ubercab.presidio.exo/secondary-dex/secondary-33af3484b3dc1dc29030b4e5680412ae7202d670.dex.jar ; echo -n :$?' failed with code 1.  Output:
53616f791d97484537820a0d445f3965Jun 07, 2017 5:43:01 PM com.facebook.buck.android.agent.AgentMain main
SEVERE: Command failed
java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:151)
	at java.net.SocketInputStream.read(SocketInputStream.java:120)
	at java.net.SocketInputStream.read(SocketInputStream.java:106)
	at com.facebook.buck.android.agent.AgentMain.receiveAndValidateSessionKey(AgentMain.java:186)
	at com.facebook.buck.android.agent.AgentMain.doReceiveFile(AgentMain.java:147)
	at com.facebook.buck.android.agent.AgentMain.main(AgentMain.java:70)


	at com.facebook.buck.android.exopackage.ExopackageInstaller$SingleDeviceInstaller.lambda$1(ExopackageInstaller.java:483)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$SingleDeviceInstaller$$Lambda$481/1478461958.accept(Unknown Source)
	at java.util.Map.forEach(Map.java:630)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$SingleDeviceInstaller.installFiles(ExopackageInstaller.java:475)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$SingleDeviceInstaller.installSecondaryDexFiles(ExopackageInstaller.java:294)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$SingleDeviceInstaller.doInstall(ExopackageInstaller.java:247)
	at com.facebook.buck.android.exopackage.ExopackageInstaller.lambda$0(ExopackageInstaller.java:206)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$$Lambda$475/1033197218.apply(Unknown Source)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$RealAdbInterface$1.call(ExopackageInstaller.java:143)
	at com.facebook.buck.android.AdbHelper$AdbCallable$1.call(AdbHelper.java:439)
	at com.facebook.buck.android.AdbHelper$AdbCallable$1.call(AdbHelper.java:1)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:111)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:58)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:75)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.facebook.buck.android.AdbHelper$CommandFailedException: Command 'umask 022 && dalvikvm -classpath /data/app/com.facebook.buck.android.agent-1.apk:/data/app/com.facebook.buck.android.agent-2.apk:/data/app/com.facebook.buck.android.agent-1/base.apk:/data/app/com.facebook.buck.android.agent-2/base.apk com.facebook.buck.android.agent.AgentMain receive-file 2828 3296974 /data/local/tmp/exopackage/com.ubercab.presidio.exo/secondary-dex/secondary-33af3484b3dc1dc29030b4e5680412ae7202d670.dex.jar ; echo -n :$?' failed with code 1.  Output:
53616f791d97484537820a0d445f3965Jun 07, 2017 5:43:01 PM com.facebook.buck.android.agent.AgentMain main
SEVERE: Command failed
java.net.SocketTimeoutException: Read timed out
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:151)
	at java.net.SocketInputStream.read(SocketInputStream.java:120)
	at java.net.SocketInputStream.read(SocketInputStream.java:106)
	at com.facebook.buck.android.agent.AgentMain.receiveAndValidateSessionKey(AgentMain.java:186)
	at com.facebook.buck.android.agent.AgentMain.doReceiveFile(AgentMain.java:147)
	at com.facebook.buck.android.agent.AgentMain.main(AgentMain.java:70)


	at com.facebook.buck.android.AdbHelper.checkReceiverOutput(AdbHelper.java:537)
	at com.facebook.buck.android.exopackage.RealExopackageDevice.installFile(RealExopackageDevice.java:261)
	at com.facebook.buck.android.exopackage.ExopackageInstaller$SingleDeviceInstaller.lambda$1(ExopackageInstaller.java:481)
```

**Comment by dreiss** [comment]
Is this 100% reproducible?  We've seen it occasionally before, but never had luck tracking it down.  It looks like perhaps adb is not properly forwarding data from the device back to the host.  I think the next step would be to run `strace -f` on adb *on the device* to see if the agent is actually sending "z1\n" over stdout (right before the call to receiveAndValidateSessionKey)

**Comment by kageiit** [comment]
@dreiss This is 100% reproducible on a few engineers' devices and laptops. Will try to see if adb is sending `z1\n` over stdout to debug

**Comment by v-jizhang** [comment]
Fixed in #1374
